core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -283,38 +305,44 @@ class KafkaServer(val config: KafkaConfig, time: Time = SystemTime) extends Logg
           val controllerId = ZkUtils.getController(zkClient)
           ZkUtils.getBrokerInfo(zkClient, controllerId) match {
             case Some(broker) =>
-              if (channel == null || prevController == null || !prevController.equals(broker)) {
-                // if this is the first attempt or if the controller has changed, create a channel to the most recent
-                // controller
-                if (channel != null) {
-                  channel.disconnect()
-                }
-                channel = new BlockingChannel(broker.getBrokerEndPoint(config.interBrokerSecurityProtocol).host,
-                  broker.getBrokerEndPoint(config.interBrokerSecurityProtocol).port,
-                  BlockingChannel.UseDefaultBufferSize,
-                  BlockingChannel.UseDefaultBufferSize,
-                  config.controllerSocketTimeoutMs)
-                channel.connect()
+              // if this is the first attempt, if the controller has changed or if an exception was thrown in a previous
+              // attempt, connect to the most recent controller
+              if (ioException || broker != prevController) {
+
+                ioException = false
+
+                if (prevController != null)
+                  selector.close(connectionId(prevController))
+
+                val brokerEndPoint = broker.getBrokerEndPoint(config.interBrokerSecurityProtocol)
+                selector.connect(
+                  connectionId(broker),
+                  new InetSocketAddress(brokerEndPoint.host, brokerEndPoint.port),
+                  Selectable.USE_DEFAULT_BUFFER_SIZE,
+                  Selectable.USE_DEFAULT_BUFFER_SIZE
+                )
```

Comment:
If you look at NetworkClient, the way that it uses Selector is the following. Before NetworkClient can send any request, it has to make sure the connector is connected and the channel is ready. If you don't follow this protocol, things can get weird. For example, in SSL, after we finish the handshake, we will turn off the interest bit for write. If the interest bit for write has already been turned on by a request sent before the handshake completes, the send may never complete.

Thinking a bit more about this. It seems to implement a blocking channel on the Selector, we will need most of the connection state management logic in NetworkClient. The problem with using NetworkClient is that it's tied to Metadata refresh, which is not needed. We can either add an option to turn off the metadata part in NetworkClient and use it to implement the blocking channel. Alternatively, we can implement a BlockingSelector on top of Selector, but has to copy some of the state management logic from NetworkClient over. Not sure which one is better.